### ISSUE TITLE
Deferred conditional in value position of mapped type is not re-evaluated when sufficient information is available to pick a branch

### ISSUE THREAD

**Issue by beezee** [microsoft/TypeScript#42385]
# Bug Report

Per [the docs on conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types), deferred conditionals are expected to be resolved once sufficient context is provided (e.g. at a call-site) such that the compiler can evaluate a single branch of the conditional.

It appears that when the conditional type is in the value position of a mapped type, this does not occur.

### 🔎 Search Terms

deferred conditional types, mapped types

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (all versions on playground), and I reviewed the entire FAQ

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-dev.20210115#code/JYOwLgpgTgZghgYwgAgCpygcwmAzgHgEEA+ZAbwChlrkAbCGMALmUKpqmEwAtnWKAvhTABPAA4p0WHMgC8yANYQRAexhoM2PPjggRxYeJQAlCLgCutMPgBiyCAA9IIACa4N0sABpWpeVK0CEgBtGwBdCkMJZAAZY1t7JwhXdwCcHxifYz9kULDE5zdcgCJ6RmL8gH5Y5BZjSNBIWEQUAHkxAjtHQtTNHFJKGmQAfRgWG3ZqXDAoOrNLaxsfac4QTAMhkHMAWzmLK1sfLe2AI2gNmloofABJH1biAAo4PYXD5BviHxVXg6XkB4ASl+iwy8X+dwBxAu1BcwAQfHwAAUnmIWGRggpkKBFMo1MgkWEQe8kZiwsQBIC5KRTPtQQSDEIECoQNM6Aw+O0CKUOcUcmQBMg4O5dCIKMzWWBkJweJyOvhijLeHy5ORBcKhXpIhK2SosfIEt1kkU0mAnnA5Z1iFTZKRHmQriwLQA6K7PMDOlY+F3HQGU8Us3VQMTcXQQFyqw1JFIeLTmy22a3UoUeuEIx5692AwHawNSspgVr65CZgs5nX5jmtYOhkDh1UqGthlyPMu5yXSri8IsNhSPJVgct5zuy6sh5sNpt1lsDnMUIA)

### 💻 Code

```ts
interface Targets<A> {
    left: A
    right: A
}
type Target = keyof Targets<any>
type Result<F extends Target, A> = Targets<A>[F]

type LR<F extends Target, L, R> = [F] extends ["left"] ? L : R

interface Ops<F extends Target> {
    _f: F
    str: Result<F, string>
    num: Result<F, number>
    lr<I, O>(a: Result<F, I>, o: Result<F, O>): Result<F, LR<F, I, O>>
    dict: <P>(p: {[k in keyof P]: Result<F, P[k]>}) => Result<F, P>
}
const left: Ops<"left"> = {} as any
const right: Ops<"right"> = {} as any

const ok = <F extends Target>(at: Ops<F>) => ({lr: at.lr(at.str, at.num)})
const orphaned = <F extends Target>(at: Ops<F>) => at.dict(ok(at))

const leftOk = ok(left)
const leftOrphaned = orphaned(left)

const rightOk = ok(right)
const rightOrphaned = orphaned(right)
```

### 🙁 Actual behavior

Note that the assignments of `leftOk` and `rightOk` properly evaluate the conditional type to the branches dictated by the respective arguments passed to `ok`, with `leftOk` taking type `{lr: string}` and `rightOk` taking type `{lr: number}`.

However with the same information available (same exact arguments, just passed through one extra layer of call-stack, and a mapped type), `leftOrphaned` and `rightOrphaned` do not evaluate the conditional type, both remaining at type `{lr: LR<F, string, number>}` despite `F` being determined by the respective arguments passed to `orphaned`.

### 🙂 Expected behavior

The types of `leftOrphaned` and `rightOrphaned` should match the types of `leftOk` and `rightOk` respectively. Application of a type argument to a type parameter (in this case "left" or "right" to `F`) should apply to sub-expressions in a function body, and those applications should hold under a mapped type.

**Comment by beezee** [comment]
Glad to see your response, thanks. @RyanCavanaugh got any pointers where I might look to familiarize myself if I wanted to take a crack at this particular issue?

**Comment by RyanCavanaugh** [comment]
Without knowing the root cause yet, I would still say this is very, very likely to require expertise beyond what we're capable of documenting. *Probably* there's a type mapper that needs to get passed around somewhere and that's not happening, but I could be way off.

**Comment by beezee** [comment]
Thanks very much @weswigham for the quick turnaround on resolution. @RyanCavanaugh I wonder if per your processes this means the fix might make it in sooner, or is the milestone assignment an imposed constraint on timeline? Anything I can do to help see this fix merged and released as soon as it could be?

**Comment by RyanCavanaugh** [comment]
If the PR gets merged before the 4.2 RC date (seems likely), it'll be in 4.2. For a long-standing issue with only one report, we wouldn't apply any extra prioritization effort beyond that.
